packet/src/main/java/org/jclouds/packet/compute/strategy/CreateSshKeysThenCreateNodes.java

```diff
@@ -65,152 +66,170 @@
+      if (options.getRunScript() != null && Strings.isNullOrEmpty(options.getLoginPrivateKey())) {
+         logger.warn(">> A runScript has been configured but no SSH key has been provided."
+               + " Authentication will delegate to the ssh-agent");
+      }
 
-        Map<String, String> defaultKeys = keyGenerator.get();
+      // If there is a key configured, then make sure there is a key pair for it
+      if (!Strings.isNullOrEmpty(options.getPublicKey())) {
+         createKeyPairForPublicKeyInOptionsAndAddToSet(options, generatedSshKeyIds);
+      }
 
-        SshKey sshKey = api.sshKeyApi().create(prefix + System.getProperty("user.name"), defaultKeys.get("public"));
-        generatedSshKeyIds.add(sshKey.id());
-        logger.debug(">> keypair created! %s", sshKey);
+      Map<?, ListenableFuture<Void>> responses = super.execute(group, count, template, goodNodes, badNodes,
+            customizationResponses);
 
-        // If a private key has not been explicitly set, configure the generated one
-        if (Strings.isNullOrEmpty(options.getLoginPrivateKey())) {
-            options.overrideLoginPrivateKey(defaultKeys.get("private"));
-        }
-    }
+      // Key pairs in Packet are only required to create the devices. They
+      // aren't used anymore so it is better
+      // to delete the auto-generated key pairs at this point where we know
+      // exactly which ones have been
+      // auto-generated by jclouds.
+      registerAutoGeneratedKeyPairCleanupCallbacks(responses, generatedSshKeyIds);
 
-    private void registerAutoGeneratedKeyPairCleanupCallbacks(Map<?, ListenableFuture<Void>> responses,
-                                                              final Set<String> generatedSshKeyIds) {
-        // The Futures.allAsList fails immediately if some of the futures fail. The Futures.successfulAsList, however,
-        // returns a list containing the results or 'null' for those futures that failed. We want to wait for all them
-        // (even if they fail), so better use the latter form.
-        ListenableFuture<List<Void>> aggregatedResponses = Futures.successfulAsList(responses.values());
+      return responses;
+   }
 
-        // Key pairs must be cleaned up after all futures completed (even if some failed).
-        Futures.addCallback(aggregatedResponses, new FutureCallback<List<Void>>() {
-            @Override
-            public void onSuccess(List<Void> result) {
-                cleanupAutoGeneratedKeyPairs(generatedSshKeyIds);
-            }
+   private void createKeyPairForPublicKeyInOptionsAndAddToSet(PacketTemplateOptions options,
+         Set<String> generatedSshKeyIds) {
+      logger.debug(">> checking if the key pair already exists...");
 
+      PublicKey userKey = readPublicKey(options.getPublicKey());
+      final String fingerprint = computeFingerprint(userKey);
+      
+      synchronized (CreateSshKeysThenCreateNodes.class) {
```

Comment:
To avoid race conditions on this "getOrCreate" method. You could have apps/clients accessing the same project with different API keys (that is, two different jclouds contexts), and race conditions could appear here.